File: android/src/main/java/com/ammarahmed/rnadmob/nativeads/RNAdMobUnifiedAdQueueWrapper.java
package com.ammarahmed.rnadmob.nativeads;

import android.content.Context;
import android.os.Handler;
import android.util.Log;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.ReactContext;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableMap;
import com.google.android.gms.ads.AdListener;
import com.google.android.gms.ads.AdLoader;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.LoadAdError;
import com.google.android.gms.ads.VideoOptions;
import com.google.android.gms.ads.admanager.AdManagerAdRequest;
import com.google.android.gms.ads.formats.AdManagerAdViewOptions;
import com.google.android.gms.ads.nativead.NativeAdOptions;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;

public class RNAdMobUnifiedAdQueueWrapper {

    public String adUnitId;
    public String name;
    public Integer totalAds = 5;
    public long expirationInterval = 3600000; // in ms
    public int totalRetryCount = 10;
    public long retryDelay = 3000;
    private int retryCount = 0;
    public Boolean muted = true;
    public Boolean mediation = false;
    public List<RNAdMobUnifiedAdContainer> nativeAds;
    //AdListener attached to attachedAdListeners list,if they are waiting for load ads
    List<AdListener> attachedAdListeners = new ArrayList<>();
    ReactContext mContext;
    int loadingAdRequestCount = 0;

    VideoOptions.Builder videoOptions;
    NativeAdOptions.Builder adOptions;
    AdListener adListener;
    private AdLoader adLoader;
    private AdManagerAdRequest.Builder adRequest;
    private UnifiedNativeAdLoadedListener unifiedNativeAdLoadedListener;
    private final Handler handler = new Handler();
    AdLoader.Builder adLoaderBuilder;

    public RNAdMobUnifiedAdQueueWrapper(ReactContext context, ReadableMap config, String repository) {
        mContext = context;
        adUnitId = config.getString("adUnitId");
        name = repository;
        videoOptions = new VideoOptions.Builder();
        adRequest = new AdManagerAdRequest.Builder();
        adOptions = new NativeAdOptions.Builder();

        adListener = new AdListener() {
            @Override
            public void onAdFailedToLoad(LoadAdError adError) {
                super.onAdFailedToLoad(adError);
                if (mediation) {
                    loadingAdRequestCount--;
                } else {
                    loadingAdRequestCount = 0;
                }
                if (loadingAdRequestCount > 0) {return;}//wait until all request failed

                boolean stopPreloading = false;
                switch (adError.getCode()) {
                    case AdRequest.ERROR_CODE_INTERNAL_ERROR:
                    case AdRequest.ERROR_CODE_INVALID_REQUEST:
                        stopPreloading = true;
                }

                if (stopPreloading) {
                    WritableMap event = Arguments.createMap();
                    WritableMap error = Arguments.createMap();
                    error.putString("message", adError.getMessage());
                    error.putInt("code", adError.getCode());
                    error.putString("domain", adError.getDomain());
                    event.putMap("error", error);

                    EventEmitter.sendEvent(mContext, CacheManager.EVENT_AD_PRELOAD_ERROR + ":" + name, event);
                    notifyOnAdsLoadFailed(adError);
                    return;
                }

                if (retryCount >= totalRetryCount) {
                    WritableMap event = getDefaultEventData();
                    WritableMap error = Arguments.createMap();
                    error.putString("message", "reach maximum retry");
                    error.putInt("code", AdRequest.ERROR_CODE_INTERNAL_ERROR);
                    error.putString("domain", "");
                    event.putMap("error", error);
                    EventEmitter.sendEvent( mContext, CacheManager.EVENT_AD_PRELOAD_ERROR + ":" + name, event);
                    notifyOnAdsLoadFailed(adError);
                    return;
                }
                retryCount++;
                handler.postDelayed(new Runnable() {
                    @Override
                    public void run() {
                        fillAds();
                    }
                }, retryDelay);

            }

            @Override
            public void onAdImpression() {
                super.onAdImpression();
                EventEmitter.sendEvent(mContext, CacheManager.EVENT_AD_IMPRESSION + ":" + name, getDefaultEventData());

            }

            @Override
            public void onAdClosed() {
                super.onAdClosed();
                WritableMap map = Arguments.createMap();
                EventEmitter.sendEvent(mContext, CacheManager.EVENT_AD_CLOSED + ":" + name, getDefaultEventData());

            }

            @Override
            public void onAdOpened() {
                super.onAdOpened();
                EventEmitter.sendEvent(mContext, CacheManager.EVENT_AD_OPEN + ":" + name, getDefaultEventData());

            }

            @Override
            public void onAdClicked() {
                super.onAdClicked();
                EventEmitter.sendEvent(mContext, CacheManager.EVENT_AD_CLICKED + ":" + name, getDefaultEventData());

            }

            @Override
            public void onAdLoaded() {
                super.onAdLoaded();

                EventEmitter.sendEvent(mContext, CacheManager.EVENT_AD_PRELOAD_LOADED + ":" + name, getDefaultEventData());

                retryCount = 0;
                if (mediation) {
                    loadingAdRequestCount--;
                } else {
                    loadingAdRequestCount = 0;
                }
                notifyOnAdsLoaded();
                if (loadingAdRequestCount == 0) { //wait until all requests finish
                    fillAds();//<-try to fill up if still not full
                }
            }
        };

        setConfiguration(config);
    }

    private WritableMap getDefaultEventData() {
        WritableMap map = Arguments.createMap();
        map.putString("adUnitId", adUnitId);
        map.putString("repo", name);
        return map;
    }

    private void  notifyOnAdsLoadFailed(LoadAdError adError){
        //use Iterator to prevent concurrentModificationException in ArrayList
        AdListener[] array = attachedAdListeners.toArray(new AdListener[0]);
        for (AdListener item : array) {
            item.onAdFailedToLoad(adError);
        }
    }
    private void  notifyOnAdsLoaded(){
        //use Iterator to prevent concurrentModificationException in ArrayList
        AdListener[] array = attachedAdListeners.toArray(new AdListener[0]);
        for (AdListener item : array) {
            item.onAdLoaded();
        }
    }

    public void attachAdListener(AdListener listener) {
        attachedAdListeners.add(listener);
    }

    public void detachAdListener(AdListener listener) {
        attachedAdListeners.remove(listener);
    }

    public void setConfiguration(ReadableMap config) {
        if (config.hasKey("retryDelay")) {
            retryDelay = config.getInt("retryDelay");
        }
        if (config.hasKey("totalRetryCount")) {
            totalRetryCount = config.getInt("totalRetryCount");
        }
        if (config.hasKey("numOfAds")) {
            totalAds = config.getInt("numOfAds");
        }
        nativeAds = new ArrayList<RNAdMobUnifiedAdContainer>(totalAds);

        if (config.hasKey("mute")) {
            muted = config.getBoolean("mute");
        }
        if (config.hasKey("expirationPeriod")) {
            expirationInterval = config.getInt("expirationPeriod");
        }
        if (config.hasKey("mediationEnabled")) {
            mediation = config.getBoolean("mediationEnabled");
        }

        if (config.hasKey("adChoicesPlacement")) {
            adOptions.setAdChoicesPlacement(config.getInt("adChoicesPlacement"));
        }

        if (config.hasKey("requestNonPersonalizedAdsOnly")) {
            Utils.setRequestNonPersonalizedAdsOnly(config.getBoolean("requestNonPersonalizedAdsOnly"), adRequest);
        }

        if (config.hasKey("mediaAspectRatio")) {
            Utils.setMediaAspectRatio(config.getInt("mediaAspectRatio"), adOptions);
        }

        if (config.hasKey("swipeGestureDirection")) {
            adOptions.enableCustomClickGestureDirection(config.getInt("swipeGestureDirection"), config.hasKey("tapsAllowed") && config.getBoolean("tapsAllowed"));
        }

        Utils.setVideoOptions(config.getMap("videoOptions"), videoOptions, adOptions);
        Utils.setTargetingOptions(config.getMap("targetingOptions"), adRequest);

        unifiedNativeAdLoadedListener = new UnifiedNativeAdLoadedListener(name, nativeAds,
                totalAds, mContext);
    }

    public void fillAds() {
        try {
            int require2fill = totalAds - nativeAds.size();
            if (require2fill <= 0 || isLoading()) {
                return;
            }

            adLoaderBuilder = new AdLoader.Builder(mContext, adUnitId);
            adLoaderBuilder.withNativeAdOptions(adOptions.build());
            adLoaderBuilder.forNativeAd(unifiedNativeAdLoadedListener);
            adLoaderBuilder.withAdListener(adListener);
            adLoader = adLoaderBuilder.build();

            Log.i("AdMob repo", "require to load >" + require2fill + "< ads more");
            loadingAdRequestCount = require2fill;
            if (mediation) {
                for (int i = 0; i < require2fill; i++) {
                    adLoader.loadAd(adRequest.build());
                }
            } else {
                adLoader.loadAds(adRequest.build(), require2fill);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public RNAdMobUnifiedAdContainer getAd() {
        if (nativeAds.isEmpty()) {
            return null;
        }
        long now = System.currentTimeMillis();
        RNAdMobUnifiedAdContainer ad = null;
        Collections.sort(nativeAds, new RNAdMobUnifiedAdComparator());
        List<RNAdMobUnifiedAdContainer> discardItems = new ArrayList<>();
        for (RNAdMobUnifiedAdContainer item : nativeAds) {
            if ((now - item.loadTime) < expirationInterval) {
                ad = item;//acceptable ad found
                break;
            } else {
                if (item.references <= 0) {
                    discardItems.add(item);
                }
            }
        }
        for (RNAdMobUnifiedAdContainer item : discardItems) {
            item.unifiedNativeAd.destroy();
            nativeAds.remove(item);
        }

        fillAds();
        if (ad == null) {
            return null;
        }
        ad.showCount += 1;
        ad.references += 1;
        return ad;
    }

    public Boolean isLoading() {
        if (adLoader != null) {
            return adLoader.isLoading() || loadingAdRequestCount > 0;
        }
        return false;
    }

    public WritableMap hasAd() {
        WritableMap args = Arguments.createMap();
        args.putInt(name, nativeAds.size());
        return args;
    }
}

File: ios/RNAdMobManager/RNAdMobUnifiedAdQueueWrapper.m
//
//  RNAdMobUnifiedAdQueueWrapper.m
//  react-native-admob-native-ads
//
//  Created by Ali on 8/25/21.
//

#import <Foundation/Foundation.h>
#import "RNAdMobUnifiedAdQueueWrapper.h"
#import "UnifiedNativeAdLoadedListener.h"
#import "RNAdMobUnifiedAdContainer.h"
#import "EventEmitter.h"
#import "CacheManager.h"
@import GoogleMobileAds;
#ifdef MEDIATION_FACEBOOK
@import FacebookAdapter;
#endif

@implementation RNAdMobUnifiedAdQueueWrapper{
    GADAdLoader* adLoader;
    GAMRequest* adRequest;

    NSMutableSet<id<AdListener>>* attachedAdListeners;
    UnifiedNativeAdLoadedListener* unifiedNativeAdLoadedListener;
    GADVideoOptions* adVideoOptions;
    GADNativeAdMediaAdLoaderOptions* adMediaOptions;
    GADNativeAdViewAdOptions* adPlacementOptions;
    NSDictionary* targetingOptions;
    int loadingAdRequestCount;
    int retryDelay;
    int totalRetryCount;
    int retryCount;
}

-(instancetype)initWithConfig:(NSDictionary *)config repo:(NSString *)repo{
    if (self = [super init])  {
        self.totalAds = 5;
        self.expirationInterval = 3600000; // in ms
        self.isMediationEnabled = false;
        retryDelay = 3000;// in ms
        totalRetryCount = 10;
        retryCount = 0;
        adRequest = [GAMRequest request];
        loadingAdRequestCount = 0;
        adVideoOptions = [[GADVideoOptions alloc]init];
        adMediaOptions = [[GADNativeAdMediaAdLoaderOptions alloc] init];
        adPlacementOptions = [[GADNativeAdViewAdOptions alloc]init];

        attachedAdListeners = [[NSMutableSet<id<AdListener>> alloc] init];
    }

    //Set repository settings
    _adUnitId = [config objectForKey:@"adUnitId"] ;
    _name = repo;
    if ([config objectForKey:@"numOfAds"]){
        _totalAds = ((NSNumber *)[config objectForKey:@"numOfAds"]).intValue;
    }
    if ([config objectForKey:@"totalRetryCount"]){
        totalRetryCount = ((NSNumber *)[config objectForKey:@"totalRetryCount"]).boolValue;
    }

    if ([config objectForKey:@"retryDelay"]){
        retryDelay = ((NSNumber *)[config objectForKey:@"retryDelay"]).boolValue;
    }

    _nativeAds =  [[NSMutableArray<RNAdMobUnifiedAdContainer *> alloc]init];

    if ([config objectForKey:@"expirationPeriod"]){
        _expirationInterval = ((NSNumber *)[config objectForKey:@"expirationPeriod"]).intValue;
    }
    if ([config objectForKey:@"mediationEnabled"]){
        _isMediationEnabled = ((NSNumber *)[config objectForKey:@"mediationEnabled"]).boolValue;
    }


    //Set request options
    if ([config objectForKey:@"adChoicesPlacement"]){
        [adPlacementOptions setPreferredAdChoicesPosition:((NSNumber *)[config objectForKey:@"adChoicesPlacement"]).intValue];
    }
    if ([config objectForKey:@"mediaAspectRatio"]){
        [adMediaOptions setMediaAspectRatio:((NSNumber *)[config objectForKey:@"mediaAspectRatio"]).intValue];
    }

    if ([config objectForKey:@"videoOptions"]){
        [self configVideoOptions:[config objectForKey:@"videoOptions"]];
    }
    if ([config objectForKey:@"mediationOptions"]){
        [self configMediationOptions:[config objectForKey:@"mediationOptions"]];
    }
    if ([config objectForKey:@"targetingOptions"]){
        [self configTargetOptions:[config objectForKey:@"targetingOptions"]];
    }


    if ([config objectForKey:@"requestNonPersonalizedAdsOnly"]){
        GADCustomEventExtras *extras = [[GADCustomEventExtras alloc] init];
        bool npa = ((NSNumber *)[config objectForKey:@"requestNonPersonalizedAdsOnly"]).boolValue;
        [extras setExtras:@{@"npa": @([NSNumber numberWithInt:npa].intValue)} forLabel:@"npa"];
        [adRequest registerAdNetworkExtras:extras];
    }

    unifiedNativeAdLoadedListener = [[UnifiedNativeAdLoadedListener alloc]initWithRepo:repo nativeAds:_nativeAds tAds:_totalAds];
    return self;
}

-(void) attachAdListener:(id<AdListener>) listener {
    [attachedAdListeners addObject:listener];
}
-(void) detachAdListener:(id<AdListener>) listener{
    [attachedAdListeners removeObject:listener];
}

/* fill up repository if need. max for multi ads request is 5 base of google admob doc
 if use mediation,you can't use GADMultipleAdsAdLoaderOptions for load ads
 */
-(void) fillAds{

    int require2fill = _totalAds-((int)_nativeAds.count);

    if ( [self isLoading] || require2fill<=0){
        return;
    }
    NSMutableArray<GADAdLoaderOptions *>* options = [NSMutableArray arrayWithArray:@[adMediaOptions,adVideoOptions,adPlacementOptions]];

    if (!_isMediationEnabled) {
        GADMultipleAdsAdLoaderOptions* multipleAdsOptions = [[GADMultipleAdsAdLoaderOptions alloc] init];
        multipleAdsOptions.numberOfAds = MAX(require2fill,0);
        [options addObject:multipleAdsOptions];
    }
    adLoader = [[GADAdLoader alloc] initWithAdUnitID:_adUnitId rootViewController:nil adTypes:@[GADAdLoaderAdTypeNative] options:options];
    [adLoader setDelegate:self];

    loadingAdRequestCount = require2fill;
    if(_isMediationEnabled){
        printf("admob request count:",MIN(require2fill,5));
        for (int i = 0; i <  MIN(require2fill,5); i++)
        {
            [adLoader loadRequest:adRequest];
        }
    }else{
        [adLoader loadRequest:adRequest];
    }
}
-(RNAdMobUnifiedAdContainer*) getAd{
    long long now = (long long)([[NSDate date] timeIntervalSince1970] * 1000.0);
    RNAdMobUnifiedAdContainer *ad = nil;

    if (!(_nativeAds.count == 0)){
        //sortAds
        [_nativeAds sortUsingComparator:^NSComparisonResult(id<Comparable, NSObject>  _Nonnull obj1,
                                                            id<Comparable, NSObject>  _Nonnull obj2) {
            return [obj1 compareTo:obj2] > 0; //find lowest showCount
        }];

        NSMutableArray<RNAdMobUnifiedAdContainer *> *discardedItems = [NSMutableArray<RNAdMobUnifiedAdContainer *> array];
        for (RNAdMobUnifiedAdContainer *item in self.nativeAds) {
            if (item != nil && (now - item.loadTime) < _expirationInterval) {
                ad = item;//acceptable ad found
                break;
            }else{
                if (item.references <=0){
                    item.unifiedNativeAd = nil;
                    [discardedItems addObject:item];
                }
            }
        }
        [self.nativeAds removeObjectsInArray:discardedItems];
    }else{
        return nil;
    }
    ad.showCount += 1;
    ad.references += 1;
    [self fillAds];
    return ad;
}
-(BOOL) isLoading{
    if (adLoader != nil){
        return [adLoader isLoading] || loadingAdRequestCount>0;
    }
    return false;
}
-(NSDictionary*) hasAd{
    NSMutableDictionary*  args = [[NSMutableDictionary alloc] init];
    [args setObject:[NSNumber numberWithInteger:_nativeAds.count] forKey:_name];
    return args;
}
- (void)adLoader:(nonnull GADAdLoader *)adLoader didReceiveNativeAd:(nonnull GADNativeAd *)nativeAd {
    loadingAdRequestCount--;
    retryCount = 0;
    [unifiedNativeAdLoadedListener adLoader:adLoader didReceiveNativeAd:nativeAd];

    //to prevent a crash (check the link below), first copy attachedAdListeners into a new array
    //link:https://stackoverflow.com/questions/44648610/collection-nsarraym-was-mutated-while-being-enumerated
    for (id<AdListener> listener in [attachedAdListeners copy]){
        [listener didAdLoaded:nativeAd];
    }

    if (loadingAdRequestCount == 0){
        [self fillAds];//fill up repository if need
    }
    // The adLoader has finished loading ads, and a new request can be sent.
}


- (void)adLoader:(nonnull GADAdLoader *)adLoader didFailToReceiveAdWithError:(nonnull NSError *)error {
      if(_isMediationEnabled){
         loadingAdRequestCount--;
      }else{
         loadingAdRequestCount = 0;
      }

    [unifiedNativeAdLoadedListener adLoader:adLoader didFailToReceiveAdWithError:error];


    BOOL stopPreloading = false;
    switch (error.code) {
        case GADErrorInternalError:
        case GADErrorInvalidRequest:
            stopPreloading = true;
            break;
    }

    if (stopPreloading) {
        NSDictionary *errorDic = @{
                @"domain":error.domain,
                @"message":error.localizedDescription,
                @"code":@(error.code).stringValue
            };
            NSDictionary *event = @{
                @"error":errorDic,
            };

            [EventEmitter.sharedInstance sendEvent:CacheManager.EVENT_AD_PRELOAD_ERROR dict:event];
            for (id<AdListener> listener in [attachedAdListeners copy]){
               [listener didFailToReceiveAdWithError:error];
             }
            return;
    }

    if (retryCount >= totalRetryCount){
        NSDictionary *errorDic = @{
               @"domain":@"",
               @"message":@"reach max retry",
               @"code":@""
            };
            NSDictionary *event = @{
                @"error":errorDic,
            };
        [EventEmitter.sharedInstance sendEvent:CacheManager.EVENT_AD_PRELOAD_ERROR dict:event];
        for (id<AdListener> listener in [attachedAdListeners copy]){
           [listener didFailToReceiveAdWithError:error];
         }
        return;
    }


    if (loadingAdRequestCount == 0) {
        retryCount++;
        __weak typeof(self) weakSelf = self;
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)((retryDelay/1000) * NSEC_PER_SEC)),
                       dispatch_get_main_queue(), ^{
            [weakSelf fillAds];
        });
    }

}

-(void)configVideoOptions:(NSDictionary *)config{

    if ([config objectForKey:@"mute"]) {
        bool muted = ((NSNumber *)[config objectForKey:@"mute"]).boolValue;
        [adVideoOptions setStartMuted:muted];
    }

    if ([config objectForKey:@"clickToExpand"]) {
        bool clickToExpand = ((NSNumber *)[config objectForKey:@"clickToExpand"]).boolValue;
        [adVideoOptions setClickToExpandRequested:clickToExpand];
    }

    if ([config objectForKey:@"customControlsRequested"]) {
        bool customControlsRequested = ((NSNumber *)[config objectForKey:@"customControlsRequested"]).boolValue;
        [adVideoOptions setCustomControlsRequested:customControlsRequested];
    }
}

-(void)configTargetOptions:(NSDictionary *)config{

    if ([config objectForKey:@"targets"]){
        NSArray<NSDictionary *>* targets = (NSArray<NSDictionary *> *)[config objectForKey:@"targets"];
        for (NSDictionary* target in targets){
             [adRequest setCustomTargeting:target];
        }

        if ([config objectForKey:@"categoryExclusions"]){
            [adRequest setCategoryExclusions:(NSArray<NSString *> *)[config objectForKey:@"categoryExclusions"]];
        }
        if ([config objectForKey:@"publisherId"]){
            [adRequest setPublisherProvidedID:(NSString *)[config objectForKey:@"publisherId"]];
        }

        if ([config objectForKey:@"requestAgent"]){
            [adRequest setRequestAgent:(NSString *)[config objectForKey:@"requestAgent"]];
        }
        if ([config objectForKey:@"keywords"]){
            [adRequest setKeywords:(NSArray<NSString *> *)[config objectForKey:@"requestAgent"]];
        }
        if ([config objectForKey:@"contentUrl"]){
            [adRequest setContentURL:(NSString *)[config objectForKey:@"contentUrl"]];
        }
        if ([config objectForKey:@"neighboringContentUrls"]){
            [adRequest setNeighboringContentURLStrings:(NSArray<NSString *> *)[config objectForKey:@"neighboringContentUrls"]];
        }

    }
}

-(void)configMediationOptions:(NSDictionary *)config{
#ifdef MEDIATION_FACEBOOK
        GADFBNetworkExtras * extras = [[GADFBNetworkExtras alloc] init];

        if ([config valueForKey:@"nativeBanner"]) {
            extras.nativeAdFormat = GADFBAdFormatNative;
        } else {
            extras.nativeAdFormat = GADFBAdFormatNativeBanner;
        }

        [adRequest registerAdNetworkExtras:extras];
#endif

}

@end
